lib/spack/spack/spec.py

```diff
@@ -1532,16 +1532,7 @@ def package_class(self):
 
     @property
     def virtual(self):
-        """Right now, a spec is virtual if no package exists with its name.
-
-        TODO: revisit this -- might need to use a separate namespace and
-        be more explicit about this.
-        Possible idea: just use conventin and make virtual deps all
-        caps, e.g., MPI vs mpi.
-        """
-        # This method can be called while regenerating the provider index
-        # So we turn off using the index to detect virtuals
-        return spack.repo.path.is_virtual(self.name, use_index=False)
+        return spack.repo.path.is_virtual(self.name)
```

Comment:
@trws I didn't look at traces in this PR, but I would expect separating the function into two optimizes times. The timings I posted seem to show a consistent shave of 0,5 sec. on the setup phase across many specs (on my laptop with SSD). 

Note that before we were using `use_index=False` because we called `Spec.virtual` even when reconstructing repo caches. In this PR I took care of using `RepoPath.is_virtual_safe` in those places so I could enable using the cache when calling `Spec.virtual`.

Reply:
Don't think it's needed, but to be more explicit, the `use_index` parameter was removed in this PR and old `use_index=False` maps to `is_virtual_safe`, while `use_index=True` maps to `is_virtual`.

Reply:
Fair, and if we want to go further in spec we could always do something like:

```python
if currently_bootstrapping():
  is_virtual_safe()
else:
  is_virtual()
```

Works for me, thanks!

Reply:
I have a last question on this one: is this going to regress https://github.com/spack/spack/pull/23661?  The point of that PR was to avoid the massive stat we do on all packages for simple operations like `spack load`, where it is faster just to check each package individually.  I wonder if doing this *looks* fast here but also slows down operations like that.  What we want to avoid is the index check for simple operations where we don't *have* to reindex all virtuals.

@alalazo can you check on this and look at the issue that it fixes?